File: backend/src/routes/commits.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { CommitsQuerySchema, CommitsByDateQuerySchema } from '../schemas/api.schema.js';
import { CommitService } from '../services/CommitService.js';
import { cacheService } from '../services/CacheService.js';

const app = new Hono();
const commitService = new CommitService();

// 获取全部已入库提交的时间范围
app.get('/date-range', async (c) => {
  const result = await cacheService.getOrSet('commits', { endpoint: 'date-range' }, 300, () =>
    commitService.getCommitDateRange()
  );

  return c.json(result);
});

// 获取提交记录列表（按日期分组）
app.get('/by-date', zValidator('query', CommitsByDateQuerySchema), async (c) => {
  const query = c.req.valid('query');


  const params = {
    startDate: query.startDate,
    endDate: query.endDate,
    repositoryIds: query.repositoryIds,
    authorEmails: query.authorEmails,
    overtimeMode: query.overtimeMode ?? (
      query.isOvertime === true
        ? 'overtime_days'
        : query.isOvertime === false
          ? 'non_overtime_days'
          : 'all'
    )
  };
  const result = await cacheService.getOrSet('commits', { endpoint: 'by-date', ...params }, 180, () =>
    commitService.getCommitsByDate(params)
  );


  return c.json({
    data: result.data,
    total: result.total,
    metricsConfig: result.metricsConfig
  });
});

// 获取提交记录列表（旧接口，保持兼容）
app.get('/', zValidator('query', CommitsQuerySchema), async (c) => {
  const query = c.req.valid('query');

  console.log('[Commits API] Query params:', {
    startDate: query.startDate,
    endDate: query.endDate,
    repositoryIds: query.repositoryIds,
    authorEmails: query.authorEmails,
    page: query.page,
    pageSize: query.pageSize
  });

  const params = {
    startDate: query.startDate,
    endDate: query.endDate,
    repositoryIds: query.repositoryIds,
    authorEmails: query.authorEmails,
    page: query.page,
    pageSize: query.pageSize
  };
  const result = await cacheService.getOrSet('commits', { endpoint: 'list', ...params }, 120, () =>
    commitService.getCommits(params)
  );

  console.log('[Commits API] Result:', {
    total: result.total,
    dataCount: result.data.length
  });

  return c.json({
    data: result.data,
    total: result.total,
    page: query.page,
    pageSize: query.pageSize
  });
});

export default app;

File: python-backend/app/api/commits.py
"""Commit API routes: query commits by date or paginated."""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.utils import parse_comma_list
from app.core.database import get_db

commits_router = APIRouter()


@commits_router.get("/by-date")
async def get_commits_by_date(
    startDate: int = Query(..., description="Start date in milliseconds"),
    endDate: int = Query(..., description="End date in milliseconds"),
    repositoryIds: str | None = Query(None, description="Comma-separated repository IDs"),
    authorEmails: str | None = Query(None, description="Comma-separated author emails"),
    isOvertime: str | None = Query("false", description="Overtime flag 'true'/'false'"),
    db: AsyncSession = Depends(get_db),
):
    """Get commits grouped by date."""
    from app.services.commit_service import CommitService

    svc = CommitService()

    repo_ids = parse_comma_list(repositoryIds)
    author_emails = parse_comma_list(authorEmails)
    overtime_flag = isOvertime == "true" if isOvertime else False

    result = await svc.get_commits_by_date(
        db,
        start_date=startDate,
        end_date=endDate,
        repository_ids=repo_ids,
        author_emails=author_emails,
        is_overtime=overtime_flag,
    )

    return {
        "data": result.get("data", []),
        "total": result.get("total", 0),
    }


@commits_router.get("")
async def get_commits(
    startDate: int = Query(..., description="Start date in milliseconds"),
    endDate: int = Query(..., description="End date in milliseconds"),
    repositoryIds: str | None = Query(None, description="Comma-separated repository IDs"),
    authorEmails: str | None = Query(None, description="Comma-separated author emails"),
    page: int = Query(1, ge=1, description="Page number"),
    pageSize: int = Query(20, ge=1, le=200, description="Page size"),
    db: AsyncSession = Depends(get_db),
):
    """Get commits with pagination."""
    from app.services.commit_service import CommitService

    svc = CommitService()

    repo_ids = parse_comma_list(repositoryIds)
    author_emails = parse_comma_list(authorEmails)

    result = await svc.get_commits(
        db,
        start_date=startDate,
        end_date=endDate,
        repository_ids=repo_ids,
        author_emails=author_emails,
        page=page,
        page_size=pageSize,
    )

    return {
        "data": result.get("data", []),
        "total": result.get("total", 0),
        "page": page,
        "pageSize": pageSize,
    }
